tst/testrandom.g

```diff
@@ -49,45 +72,79 @@ randomTest := function(collection, method, checkin...)
-        return;
-    fi;
+# Here we can't check different seeds produce different answers
+# We do check that the random source is not used, for efficency.
+randomTestForSizeOneCollectionInner := function(filter, global_rs, global_randfunc, randfunc, collection, checkin)
+    local i, val, local_rs, intlist1, intlist2;
 
     val := Representative(collection);
 
-    Init(GlobalMersenneTwister, 6);
-    intlist1 := List([1..1000], x -> Random([1..10]));
+    Init(global_rs, 6);
+    intlist1 := List([1..10], x -> global_randfunc([1..1000]));
 
-    for i in [1..1000] do
-        if method(collection) <> val then
+    for i in [1..100] do
+        if global_randfunc(collection) <> val then
             Print("Random returned something outside collection :", collection, ":", val);
         fi;
     od;
 
-    for i in [1..1000] do
-        if method(GlobalMersenneTwister, collection) <> val then
+    for i in [1..100] do
+        if randfunc(global_rs, collection) <> val then
             Print("Random returned something outside collection :", collection, ":", val);
         fi;
     od;
 
-    localgen := RandomSource(IsMersenneTwister, 6);
+    local_rs := RandomSource(filter, 6);
 
-    Init(GlobalMersenneTwister, 6);
-    for i in [1..1000] do
-        if method(localgen, collection) <> val then
+    Init(global_rs, 6);
+    for i in [1..100] do
+        if randfunc(local_rs, collection) <> val then
             Print("Random returned something outside collection :", collection, ":", val);
         fi;
     od;
 
-    # The previous loop should not have affected GlobalMersenneTwister,
+    # The previous loop should not have affected global_rs,
     # so this should be the same as intlist1
-    intlist2 := List([1..1000], x -> Random([1..10]));
+    intlist2 := List([1..10], x -> global_randfunc([1..1000]));
 
     if intlist1 <> intlist2 then
         Print("Random read from local gen affected global gen: ", collection);
     fi;
 end;;
+
+
+randomTest := function(collection, randfunc, checkin...)
+    local sizeone;
+    if Length(checkin) = 0 then
+        checkin := \in;
+    else
+        checkin := checkin[1];
+    fi;
+
+    # Make a best attempt to find if the collection is size 1.
+    # There are implementations of random for objects which do not support
+    # Size or IsTrivial, e.g. PadicExtensionNumberFamily
+    if IsList(collection) then
+        sizeone := (Size(collection) = 1);
+    elif IsCollection(collection) then
+        sizeone := IsTrivial(collection);
+    else
+        sizeone := false;
+    fi;
+
+    if sizeone then
+        randomTestForSizeOneCollectionInner(IsMersenneTwister, 
+          GlobalMersenneTwister, x -> randfunc(x), randfunc, collection, checkin);
+        randomTestForSizeOneCollectionInner(IsGAPRandomSource, 
+          GlobalRandomSource, x -> randfunc(GlobalRandomSource, x), randfunc,
+          collection, checkin);
+    else
+        randomTestInner(IsMersenneTwister,
```

Comment:
You could avoid some code duplication by adding a local var which you set to either ˋrandomTestInnerˋ or the orher func.